README.md

```diff
@@ -82,6 +83,60 @@ If using Rails, this is evaluated against the timezone configured in Rails, othe
 If you want to have your jobs enqueued based on a different time zone you can specify a timezone in the cronline,
 like this `'0 22 * * 1-5 America/Chicago'`.
 
+### Namespacing
+
+#### Default namespace
+
+When not giving a namespace, the `default` one will be used.
+
+In the case you'd like to change this value, create a new initializer like so:
+
+`config/initializers/sidekiq-cron.rb`:
+
+```ruby
+Sidekiq::Cron.configure do |config|
+  config.default_namespace = 'statics'
+end
+```
+
+#### Usage
+
+When creating a new job, you can optionaly give a `namespace` attribute, and then you can pass it too in the `find` or `destroy` methods.
+
+```ruby
+Sidekiq::Cron::Job.create(
+  name: 'Hard worker - every 5min',
+  namespace: 'Foo',
+  cron: '*/5 * * * *',
+  class: 'HardWorker'
+)
+# INFO: Cron Jobs - add job with name Hard worker - every 5min in the namespace Foo
+
+# Without specifing the namespace, Sidekiq::Cron use the `default` one, therefore `count` return 0.
+Sidekiq::Cron::Job.count
+#=> 0
+
+# Searching in the job's namespace returns 1.
+Sidekiq::Cron::Job.count 'Foo'
+#=> 1
+
+# Same applies to `all`. Without a namespace, no jobs found.
+Sidekiq::Cron::Job.all
+
+# But giving the job's namespace returns it.
+Sidekiq::Cron::Job.all 'Foo'
+#=> [#<Sidekiq::Cron::Job:0x00007f7848a326a0 ... @name="Hard worker - every 5min", @namespace="Foo", @cron="*/5 * * * *", @klass="HardWorker", @status="enabled" ... >]
+
+# If you'd like to get all the jobs across all the namespaces then pass an asterisk:
+Sidekiq::Cron::Job.all '*'
```

Comment:
I am not sure about this API. Why does `Sidekiq::Cron::Job.all` not return all jobs in all namespaces?

Basically it would seem more logical to me that not specifying a namespace would search all jobs.

Reply:
I made `Sidekiq::Cron::Job.all` returning the jobs from the `default_namespace` allowing one to fetch jobs only from that namespace (like Kubernetes does for example), and when you want really ALL the jobs, then you pass the asterisk symbol.

That look logical to me, but if you really want `Sidekiq::Cron::Job.all` to return ALL jobs by default, then you need to clarify when is the `default_namespace` used by the user please.

Reply:
I am not sure I understand your comment.

`Sidekiq::Cron::Job` methods `all`, `count` and others should return all jobs regardless of namespace
user can call `Sidekiq::Cron::Job.all('default')` to get the jobs from the default namespace

Reply:
The goal of namespaces here is to segregate jobs in groups so that one can execute actions on all the jobs from a specific namespace, and be sure that other jobs from other namespaces are untouched.

Forcing the dev to tell the namespace from which to execute that action, or forcing the dev to retrieve all the jobs by passing an asterisk prevent him from doing a mistake and ending by deactivating, or worst destroying all the jobs from all the namespaces.

At the end of the day, for all the devs using this gem, when they don't want to use namespaces, they just have to ignore it and the gem will still work in the same way : all the jobs will go to the default namespace specified by the `default_namespace` configuration option (which is `default` by default), and using the `Sidekiq::Cron::Job` methods will return all their jobs without passing any argument (see all your tests are still passing).

From our point of view, this is a nice trade-off :
* people which doesn't care about namespaces will use exactly the same way this gem
* people using namespaces will tell what they want (a specific namespace, or no namespaces and get all the jobs across all the existing namespaces) so that they can't make mistakes.

Reply:
I agree with you, but: 
What concerns me is that Job.all(nil) has undefined behaviour in the current code. 
My thinking is that you either need to add a check for nil into jobs_key() and raise an error when nil is passed
or make nil represent the "*". Which to me looks like to cleaner option because it means less code (no special handling for nil and "*", no method arg default value reading from config) and that means less complexity a cleaner API. 
I think that nobody will be surprised that Job.all will return all jobs from all namespaces.
Another solution that comes to mind is to remove this arg from Job completely and have it always return everything (just like before this PR). And use a different API to work with namespaced Jobs like `Sidekiq::Cron::Namespace.new('default').jobs`

Reply:
I see, that's a good catch. I just pushed a fix for that case with more unit test testing with `nil`, `""` and when namespace doesn't exist.

Could you please have a look and give us your feedback ?

Reply:
Thanks for the quick fix, however I don't agree with the direction you are taking. 
@markets WDYT?

Reply:
My team does not plan to use `namespaces` for now, so I 100% agree with that point:

> At the end of the day, for all the devs using this gem, when they don't want to use namespaces, they just have to ignore it and the gem will still work in the same way

So I think the trade-off presented by @zedtux makes sense: for people not using `namespaces` this should be a "transparent" upgrade with no risk, and then if someone wants/needs to use `namespaces` they should pass the namespace in all operations.

Reply:
@markets yes, we are all in agreement in that, the question is about the API design for getting all jobs.

`Sidekiq::Cron::Job.all('*')`  vs `Sidekiq::Cron::Job.all`

Reply:
What about the following API?
- `Sidekiq::Cron::Job.all('*')` => returns all jobs from all namespaces
- `Sidekiq::Cron::Job.all` => returns all jobs from the `default_namespace`
- `Sidekiq::Cron::Job.all(namespace)` => returns all jobs from the given `namespace`

That way, for people not using `namespaces` functionality, they can continue using same API with no changes (`Sidekiq::Cron::Job.all`).

Reply:
that is how it is implemented today, hence I will close do a final review and merge

Reply:
One last point coming to my mind, is "migration". I will test the namespace code with a job without a namespace in order to avoid breaking existing jobs when migrating to the newer version of the gem.

Reply:
Yep, we should ensure "migration" is gonna be smooth for all current installations and people can just upgrade the gem and everything is working as usual.

Reply:
Ok, one final argument and then I will give up on this debate:
Currently my code uses Sidekiq::Cron::Job.all to do something all the Cron jobs scheduled. Once I switch to namespaces for some component of mine this call will suddenly change behaviour because it will no longer return all the jobs butly only the ones in `default` namespace. This seems like a unnecessary breaking change.
What do you think @markets ?

Reply:
@honzasterba @zedtux Even I like the concept of having the ability to define "namespaces", I'm starting to think that this change will be hard to release without any kind of breaking change... Maybe we should release it as a separate plugin for now? Something like `sidekiq-cron-namespaces`... So people can decide if they want to use the feature and "fight" with potential problems during the migration process.

As I said before, I like this idea... but being honest, the migration seems a bit "dangerous" and there is a lot of changes (in critical paths) to review in that branch that I'm not confident about...

Reply:
@zedtux ok! we'll need to perform a more in depth review (sorry 🙏🏼 I didn't touched this part of the codebase and I'd be more confident if other maintainers/contributors also review this branch), but +1 to ship this in future releases. 

I just made a quick re-review and it seems we are on the right path now (especially regarding the auto-migration process). I'll test this out in my apps.

But, I think we should focus first on finish the Sidekiq v7 support, which is happening on #369 and touches also some internals.

Thank you so much for your work!

Reply:
@markets I just realise now you've done the Sidekiq v7 support. The awesome bot is starting to bother on this PR now ... and it's the new year, so what's the plan ? Rebase / retest / merge ? 😎

Reply:
Hi @zedtux 👋🏼 

Yeah, I think we are ready to continue with this branch. IMO next steps should be: rebase, finish review (I'd like to hear final thoughts from @honzasterba and/or @ondrejbartas too), test and merge 🚀

Reply:
Do you have any news @markets please?

Reply:
I'm overall ok with the feature, but we still need to resolve some @honzasterba's comments (and update from master, we added quite new things and fixes lately).

Reply:
Okay, I'll go through the comments one more time, hopefully the last time. Thanks.